File: openclaw_republic/government.py
"""CyberGovernment — 三权分立 AI 协作政府的主入口。

负责初始化三权分支的所有 Agent、创建并启动消息总线、
管理法案生命周期、提供外部 API（receive_petition）。
"""

from __future__ import annotations

import logging
import uuid
from pathlib import Path
from typing import Any

from openclaw_republic.agents.executive.engine import ExecutionEngine, TaskExecutor
from openclaw_republic.agents.executive.president import President
from openclaw_republic.agents.executive.sec_engineering import SecretaryOfEngineering
from openclaw_republic.agents.executive.sec_state import SecretaryOfState
from openclaw_republic.agents.judicial.chief_justice import ChiefJustice
from openclaw_republic.agents.legislative.conservative_mp import ConservativeMP
from openclaw_republic.agents.legislative.debate import VoteRecord, VoteResult
from openclaw_republic.agents.legislative.radical_mp import RadicalMP
from openclaw_republic.agents.legislative.speaker import Speaker
from openclaw_republic.bus.event_log import EventLogger
from openclaw_republic.bus.message_bus import MessageBus
from openclaw_republic.bus.state_machine import BillLifecycle, BillState
from openclaw_republic.config.loader import load_constitution
from openclaw_republic.config.models import ConstitutionConfig
from openclaw_republic.schemas.events import (
    BaseEvent,
    DebateEvent,
    EventAction,
    ExecutionEvent,
    VetoEvent,
    VoteEvent,
)

logger = logging.getLogger(__name__)

# 最大重试次数（防止 Veto/违宪无限回路），演示用设为 1
_MAX_RETRIES = 1


class CyberGovernment:
    """三权分立 AI 协作政府的主入口。

    负责：
    1. 初始化三权分支的所有 Agent
    2. 创建并启动消息总线
    3. 注册各分支到消息总线
    4. 管理法案生命周期
    5. 提供外部 API（接收选民请愿）
    """

    def __init__(
        self,
        config_dir: Path | str,
        *,
        constitution: ConstitutionConfig | None = None,
    ) -> None:
        """初始化三权协作政府。

        Args:
            config_dir: 配置文件目录（含 constitution.yaml）。
            constitution: 直接传入宪法配置（跳过文件加载，用于测试）。
        """
        self._config_dir = Path(config_dir)

        # 加载宪法
        if constitution is not None:
            self.constitution = constitution
        else:
            self.constitution = load_constitution(
                self._config_dir / "constitution.yaml",
            )

        # 初始化各分支 Agent
        self._init_legislative()
        self._init_executive()
        self._init_judicial()

        # 消息总线 & 事件日志
        self.bus = MessageBus()
        self.event_logger = EventLogger()
        self._register_subscribers()

    # ─── 分支初始化 ─────────────────────────

    def _init_legislative(self) -> None:
        """初始化立法分支 Agent。"""
        self.speaker = Speaker()
        self.radical_mp = RadicalMP()
        self.conservative_mp = ConservativeMP()

    def _init_executive(self) -> None:
        """初始化行政分支 Agent。"""
        budget = self.constitution.judicial.token_budget.execution_budget
        self.president = President(token_budget=budget)
        self.sec_engineering = SecretaryOfEngineering()
        self.sec_state = SecretaryOfState()

        # 组建内阁 → 执行引擎
        cabinet: dict[str, TaskExecutor] = {
            "CodeExecution": self.sec_engineering,
            "Python_Interpreter": self.sec_engineering,
            "GitHub": self.sec_engineering,
            "WebBrowser": self.sec_state,
            "Search": self.sec_state,
        }
        self.execution_engine = ExecutionEngine(cabinet)

    def _init_judicial(self) -> None:
        """初始化司法分支 Agent。"""
        self.chief_justice = ChiefJustice(self.constitution)

    # ─── 消息总线订阅 ───────────────────────

    def _register_subscribers(self) -> None:
        """注册各分支到消息总线主题。"""
        # 所有主题统一记录到事件日志
        for topic in ("legislation", "execution", "judiciary", "lifecycle"):
            self.bus.subscribe(topic, self._log_event)

    async def _log_event(self, event: BaseEvent) -> None:
        """统一事件日志处理器。"""
        self.event_logger.log(event)

    # ─── 公开接口 ─────────────────────────

    async def inaugurate(self) -> None:
        """启动三权协作系统。"""
        await self.bus.start()
        logger.info("CyberGovernment 已启动")

    async def shutdown(self) -> None:
        """关闭三权协作系统。"""
        await self.bus.stop()
        logger.info("CyberGovernment 已关闭")

    async def receive_petition(
        self,
        petition: str,
        *,
        max_retries: int = _MAX_RETRIES,
        task_id: str | None = None,
    ) -> str:
        """接收选民请愿，启动完整 Pipeline。

        Pipeline 流程（含回路重试）：
        1. 创建 BillLifecycle → DRAFTING
        2. 辩论 → DEBATING → VOTED
        3. 总统审查 → SIGNED 或 VETOED
        4. 执行引擎执行 → EXECUTING → REVIEWING
        5. 大法官审查 → CONSTITUTIONAL → DELIVERED
           或 UNCONSTITUTIONAL → 回 DRAFTING

        Args:
            petition: 选民请愿内容。
            max_retries: 最大回路重试次数（防止无限循环）。
            task_id: 外部传入的追踪 ID，用于 WebSocket 绑定对应主题。

        Returns:
            最终交付结果描述。
        """
        bill_id = task_id if task_id else str(uuid.uuid4())
        lifecycle = BillLifecycle(bill_id)

        for attempt in range(1, max_retries + 1):
            logger.info(
                "Pipeline attempt %d/%d for bill %s",
                attempt,
                max_retries,
                bill_id,
            )

            result = await self._run_pipeline(
                petition,
                lifecycle,
                bill_id,
            )

            if result is not None:
                return result

            # 回路：lifecycle 已 transition 回 DRAFTING
            logger.info("Bill %s 回到 DRAFTING，重试第 %d 次", bill_id, attempt)

        return (
            f"法案 {bill_id} 在 {max_retries} 次重试后仍未通过。"
            f"当前状态: {lifecycle.current_state.value}"
        )

    # ─── Pipeline 内部逻辑 ─────────────────

    async def _run_pipeline(
        self,
        petition: str,
        lifecycle: BillLifecycle,
        bill_id: str,
    ) -> str | None:
        """执行单次 Pipeline 流程。

        Returns:
            成功交付时返回结果字符串；需要回路重试时返回 None。
        """
        # 1. → DRAFTING（首次从 PETITION 转换；回路重试时已在 DRAFTING）
        if lifecycle.current_state != BillState.DRAFTING:
            lifecycle.transition(BillState.DRAFTING)
        await self._publish_lifecycle(bill_id, "drafting")

        async def debate_publisher(action: EventAction, **kwargs: Any) -> None:
            if action == EventAction.PROPOSE:
                await self._publish_propose(
                    bill_id,
                    str(kwargs.get("agent", "unknown")),
                    str(kwargs.get("text", "")),
                    str(kwargs.get("emotion", "neutral")),
                    int(kwargs.get("round_number", 1)),
                    float(kwargs.get("conflict_score", 0.0)),
                )
            elif action == EventAction.BRAWL:
                await self._publish_brawl(bill_id, float(kwargs.get("intensity", 0.5)))
            elif action == EventAction.ORDER:
                await self._publish_order(bill_id, float(kwargs.get("intensity", 0.5)))

        async def execution_publisher(status: str, skill: str, step_index: int) -> None:
            await self._publish_tool_call(bill_id, skill, step_index, status)

        # 2. DRAFTING → DEBATING
        await self.speaker.receive_petition(petition)
        lifecycle.transition(BillState.DEBATING)
        await self._publish_lifecycle(bill_id, "debating")

        debate_result = await self.speaker.moderate_debate(
            self.radical_mp,
            self.conservative_mp,
            self.constitution.judicial.debate,
            event_publisher=debate_publisher,
        )

        # 3. DEBATING → VOTED
        lifecycle.transition(BillState.VOTED)
        await self._publish_lifecycle(bill_id, "voted")

        vote_result = await self.speaker.call_vote(
            debate_result.final_proposal,
            [self.radical_mp, self.conservative_mp],
        )

        # Mock LLM 模式下 vote 可能不通过（空字符串不匹配投票关键字），
        # 此时强制通过以保证 Pipeline 可端到端运行。
        if not vote_result.passed:
            vote_result = self._force_vote_passed(vote_result)

        # 生成法案
        act = await self.speaker.generate_act(
            petition,
            debate_result,
            vote_result,
        )

        await self._publish_vote_passed(bill_id, vote_result.ayes, vote_result.nays, act=act)

        # 4. 总统审查
        veto = await self.president.review_act(act)

        if veto is not None:
            # VOTED → VETOED → DRAFTING
            lifecycle.transition(BillState.VETOED)
            await self._publish_veto(bill_id, veto.reason)
            lifecycle.transition(BillState.DRAFTING)
            return None  # 触发重试

        # VOTED → SIGNED
        lifecycle.transition(BillState.SIGNED)
        await self._publish_lifecycle(bill_id, "signed")
        await self._publish_sign(bill_id)

        # 5. SIGNED → EXECUTING
        lifecycle.transition(BillState.EXECUTING)
        await self._publish_lifecycle(bill_id, "executing")

        report = await self.execution_engine.execute_act(act, event_publisher=execution_publisher)

        # 6. EXECUTING → REVIEWING
        lifecycle.transition(BillState.REVIEWING)
        await self._publish_lifecycle(bill_id, "reviewing")

        verdict = await self.chief_justice.review_result(petition, report)
        judgment = await self.chief_justice.issue_judgment(verdict)
        judgment.payload["verdict"] = verdict.model_dump(mode="json")

        # 强制修正 judgment 事件的 task_id，避免首席大法官内部使用了随机的 act_id 导致前端收不到结果
        judgment.task_id = bill_id

        # 发布判决事件
        await self.bus.publish("judiciary", judgment)

        if not verdict.constitutional:
            # REVIEWING → UNCONSTITUTIONAL → DRAFTING
            lifecycle.transition(BillState.UNCONSTITUTIONAL)
            lifecycle.transition(BillState.DRAFTING)
            return None  # 触发重试

        # 7. REVIEWING → CONSTITUTIONAL → DELIVERED
        lifecycle.transition(BillState.CONSTITUTIONAL)
        lifecycle.transition(BillState.DELIVERED)
        await self._publish_lifecycle(bill_id, "delivered")

        return (
            f"法案 {bill_id} 已交付。\n"
            f"执行状态: {report.overall_status}\n"
            f"判决: {verdict.ruling}\n"
            f"总 Token 消耗: {report.total_tokens_consumed}"
        )

    # ─── 事件发布辅助 ─────────────────────

    async def _publish_lifecycle(self, bill_id: str, state: str) -> None:
        """发布生命周期状态变更事件。"""
        event = BaseEvent(
            source_agent="government",
            action=EventAction.STATE_CHANGE,  # 法案生命周期状态变更
            payload={"bill_id": bill_id, "state": state},
            task_id=bill_id,
        )
        await self.bus.publish("lifecycle", event)

    async def _publish_veto(self, bill_id: str, reason: str) -> None:
        """发布否决事件。"""
        event = VetoEvent(
            source_agent="president",
            reason=reason,
            task_id=bill_id,
        )
        await self.bus.publish("legislation", event)

    async def _publish_propose(
        self,
        bill_id: str,
        agent: str,
        text: str,
        emotion: str = "neutral",
        round_number: int = 1,
        conflict_score: float = 0.0,
    ) -> None:
        """发布提案/反驳事件。"""
        event = DebateEvent(
            source_agent=agent,
            action=EventAction.PROPOSE,
            emotion=emotion,  # type: ignore
            statement=text,
            task_id=bill_id,
            round_number=round_number,
            conflict_score=conflict_score,
        )
        await self.bus.publish("legislation", event)

    async def _publish_brawl(self, bill_id: str, intensity: float) -> None:
        """发布辩论激烈事件。"""
        event = BaseEvent(
            source_agent="speaker",
            action=EventAction.BRAWL,
            intensity=intensity,
            task_id=bill_id,
        )
        await self.bus.publish("legislation", event)

    async def _publish_order(self, bill_id: str, intensity: float) -> None:
        """发布议长控场事件。"""
        event = BaseEvent(
            source_agent="speaker",
            action=EventAction.ORDER,
            intensity=intensity,
            task_id=bill_id,
        )
        await self.bus.publish("legislation", event)

    async def _publish_vote_passed(
        self, bill_id: str, ayes: int, nays: int, act: Any | None = None
    ) -> None:
        """发布表决通过事件。"""
        payload = {}
        if act is not None:
            # We assume act has model_dump method (Pydantic object)
            payload["act"] = act.model_dump(mode="json")

        event = VoteEvent(
            source_agent="speaker",
            ayes=ayes,
            nays=nays,
            result="passed",
            task_id=bill_id,
            payload=payload,
        )
        await self.bus.publish("legislation", event)

    async def _publish_sign(self, bill_id: str) -> None:
        """发布总统签署事件。"""
        event = BaseEvent(
            source_agent="president",
            action=EventAction.SIGN_ACT,
            task_id=bill_id,
        )
        await self.bus.publish("legislation", event)

    async def _publish_tool_call(
        self, bill_id: str, skill: str, step_index: int, status: str
    ) -> None:
        """发布工具调用事件。"""
        event = ExecutionEvent(
            source_agent="sec_engineering",
            action=EventAction.TOOL_CALL,
            tool_name=skill,
            step_index=step_index,
            status=status,
            task_id=bill_id,
        )
        await self.bus.publish("execution", event)

    @staticmethod
    def _force_vote_passed(vote_result: VoteResult) -> VoteResult:
        """强制将投票结果设为通过。

        Mock LLM 模式下 ``_call_llm`` 返回空字符串，导致所有
        议员投票均返回 False。此方法将投票结果强制为 passed，
        保证 Pipeline 可端到端运行。

        Args:
            vote_result: 原始投票结果。

        Returns:
            修正后的投票结果（passed=True）。
        """
        return VoteResult(
            proposal=vote_result.proposal,
            records=[VoteRecord(voter_role=r.voter_role, vote=True) for r in vote_result.records],
            ayes=len(vote_result.records),
            nays=0,
            passed=True,
        )

File: backend/src/government.ts
import { randomUUID } from 'node:crypto';
import { Speaker } from './agents/legislative/speaker';
import { RadicalMP } from './agents/legislative/radical-mp';
import { ConservativeMP } from './agents/legislative/conservative-mp';
import { President } from './agents/executive/president';
import { SecretaryOfEngineering } from './agents/executive/sec-engineering';
import { SecretaryOfState } from './agents/executive/sec-state';
import { ExecutionEngine, TaskExecutor } from './agents/executive/engine';
import { ChiefJustice } from './agents/judicial/chief-justice';
import { MessageBus } from './bus/message-bus';
import { EventLogger } from './bus/event-log';
import { BillLifecycle, BillState } from './bus/state-machine';
import { loadConstitution, resolveModel } from './config/loader';
import { ConstitutionConfig } from './config/models';
import { OpenClawAdapter } from './openclaw/adapter';
import { VoteResult } from './agents/legislative/debate';
import { 
  BaseEvent, EventAction, EmotionType, VoteEvent
} from './schemas/events';
import { ViolationType } from './schemas/verdict';

const MAX_RETRIES = 1;

export class CyberGovernment {
  private _configDir: string;
  public constitution: ConstitutionConfig;
  public adapter: OpenClawAdapter;
  
  public speaker!: Speaker;
  public radicalMp!: RadicalMP;
  public conservativeMp!: ConservativeMP;
  
  public president!: President;
  public secEngineering!: SecretaryOfEngineering;
  public secState!: SecretaryOfState;
  public executionEngine!: ExecutionEngine;
  
  public chiefJustice!: ChiefJustice;
  
  public bus: MessageBus;
  public eventLogger: EventLogger;

  /** Cumulative token counters per branch */
  private _tokenCounters = { legislative: 0, executive: 0, judicial: 0 };

  constructor(configDir: string, constitution?: ConstitutionConfig) {
    this._configDir = configDir;
    
    if (constitution) {
      this.constitution = constitution;
    } else {
      this.constitution = loadConstitution(this._configDir);
    }

    this.adapter = new OpenClawAdapter();
    this.bus = new MessageBus();
    this.eventLogger = new EventLogger();

    this._initLegislative();
    this._initExecutive();
    this._initJudicial();
    this._applyModelRouting();
    this._registerSubscribers();
  }

  private _initLegislative(): void {
    this.speaker = new Speaker(this.adapter, this.bus);
    this.radicalMp = new RadicalMP(this.adapter, this.bus);
    this.conservativeMp = new ConservativeMP(this.adapter, this.bus);
  }

  private _initExecutive(): void {
    const budget = this.constitution.judicial.token_budget?.execution_budget ?? 50000;
    this.president = new President(this.adapter, this.bus, budget);
    this.secEngineering = new SecretaryOfEngineering(this.adapter, this.bus);
    this.secState = new SecretaryOfState(this.adapter, this.bus);

    const cabinet: Record<string, TaskExecutor> = {
      "CodeExecution": this.secEngineering,
      "Python_Interpreter": this.secEngineering,
      "GitHub": this.secEngineering,
      "WebBrowser": this.secState,
      "Search": this.secState,
    };
    this.executionEngine = new ExecutionEngine(cabinet);
  }

  private _initJudicial(): void {
    this.chiefJustice = new ChiefJustice(this.constitution, this.adapter, this.bus);
  }

  /**
   * Inject per-agent model overrides from constitution.model_routing.
   * Must be called after all agents are initialized.
   */
  private _applyModelRouting(): void {
    const routing = this.constitution.model_routing;
    if (!routing) return;

    const agents = [
      this.speaker, this.radicalMp, this.conservativeMp,
      this.president, this.secEngineering, this.secState,
      this.chiefJustice,
    ];

    for (const agent of agents) {
      agent.modelRef = resolveModel(agent.role, routing);
    }
  }

  private _registerSubscribers(): void {
    const topics = ["legislation", "execution", "judiciary", "lifecycle"] as const;
    for (const topic of topics) {
      this.bus.subscribe(topic, this._logEvent.bind(this));
    }
  }

  private async _logEvent(event: BaseEvent): Promise<void> {
    await this.eventLogger.log(event);
  }

  public async inaugurate(): Promise<void> {
    await this.bus.start();
    console.log("CyberGovernment 已启动");
  }

  public async shutdown(): Promise<void> {
    await this.bus.stop();
    console.log("CyberGovernment 已关闭");
  }

  /**
   * 接收选民请愿，启动完整 Pipeline。
   */
  public async receivePetition(
    petition: string, 
    maxRetries: number = MAX_RETRIES, 
    taskId?: string
  ): Promise<string> {
    const billId = taskId ?? randomUUID();
    const lifecycle = new BillLifecycle(billId);

    for (let attempt = 1; attempt <= 1 + maxRetries; attempt++) {
      console.log(`Pipeline attempt ${attempt}/${1 + maxRetries} for bill ${billId}`);

      try {
        const result = await this._runPipeline(petition, lifecycle, billId);

        if (result !== null) {
          return result;
        }
      } catch (err: any) {
        // Unhandled system-level exception — log and abort
        const message = err instanceof Error ? err.message : String(err);
        console.error(`[CyberGovernment] 系统级异常 in pipeline attempt ${attempt}: ${message}`);

        // Force lifecycle back to DRAFTING for safety
        try { lifecycle.transition(BillState.DRAFTING); } catch { /* ignore transition errors */ }

        return `系统级异常: ${message}。流水线已中止。法案 ${billId} 未完成。`;
      }

      console.log(`Bill ${billId} 回到 DRAFTING，重试第 ${attempt} 次`);
    }

    return `法案 ${billId} 在 ${maxRetries} 次重试后仍未通过。系统已拦截并终止该危险案卷。`;
  }

  private async _runPipeline(
    petition: string, 
    lifecycle: BillLifecycle, 
    billId: string
  ): Promise<string | null> {
    if (lifecycle.current_state !== BillState.DRAFTING) {
      lifecycle.transition(BillState.DRAFTING);
    }
    await this._publishLifecycle(billId, "drafting");

    // Bug 23 fix: 重置 ProcessReviewer 历史，防止跨 Act 死循环误报
    this.chiefJustice.resetProcessHistory();

    await this.speaker.receivePetition(petition);
    
    lifecycle.transition(BillState.DEBATING);
    await this._publishLifecycle(billId, "debating");

    const debateResult = await this.speaker.moderateDebate(
      this.radicalMp,
      this.conservativeMp,
      this.constitution.judicial.debate,
      billId,
      petition  // Bug 40 fix: 显式传递 petition，避免并发 pipeline 竞态
    );

    // Token 埋点①：立法辩论阶段
    // 估算：每轮辩论约 2000 tokens（提案 + 批评 + 反驳 + 议长介入）
    const debateTokens = debateResult.rounds.length * 2000;
    this._tokenCounters.legislative += debateTokens;
    await this._publishTokenUsage(billId, 'legislative', debateTokens, this._tokenCounters.legislative);

    lifecycle.transition(BillState.VOTED);
    await this._publishLifecycle(billId, "voted");

    let voteResult = await this.speaker.callVote(
      debateResult.final_proposal,
      [this.radicalMp, this.conservativeMp],
      debateResult.rounds.length + 1,
      billId
    );

    if (!voteResult.passed) {
      voteResult = this._forceVotePassed(voteResult);
    }

    const act = await this.speaker.generateAct(petition, debateResult, voteResult);
    act.act_id = billId; // Force telemetry trace linkage
    await this._publishVotePassed(billId, voteResult.ayes, voteResult.nays, act);

    const veto = await this.president.evaluateAct(act);

    // Token 埋点②：总统审查法案
    const signTokens = 1500;
    this._tokenCounters.executive += signTokens;
    await this._publishTokenUsage(billId, 'executive', signTokens, this._tokenCounters.executive);

    if (veto !== null && veto !== undefined) {
      lifecycle.transition(BillState.VETOED);
      // President emits VETO event internally
      lifecycle.transition(BillState.DRAFTING);
      return null;
    }

    lifecycle.transition(BillState.SIGNED);
    await this._publishLifecycle(billId, "signed");
    // President emits SIGN_ACT internally during evaluateAct when successful

    lifecycle.transition(BillState.EXECUTING);
    await this._publishLifecycle(billId, "executing");

    const report = await this.executionEngine.executeAct(act);

    // Token 埋点③：内阁执行阶段
    const execTokens = report.total_tokens_consumed || 0;
    this._tokenCounters.executive += execTokens;
    await this._publishTokenUsage(billId, 'executive', execTokens, this._tokenCounters.executive);

    // Bug 6+49 fix: 全失败或部分失败的执行不应送去司法审查
    // partial: 残缺产出送给 ResultReviewer 会因为只有部分成功输出而产生虚假偏离度评分
    // failed: ResultReviewer 会把 '(无有效产出)' 传给偏离度评分器，产生虚假的"合宪"判定
    if (report.overall_status === 'failed' || report.overall_status === 'partial') {
      console.warn(`[CyberGovernment] 执行状态=${report.overall_status} (bill ${billId})，触发自动违宪判决并重试`);
      lifecycle.transition(BillState.REVIEWING);
      await this._publishLifecycle(billId, "reviewing");
      
      const fakeVerdict = {
        verdict_id: randomUUID(),
        act_id: billId,
        constitutional: false,
        ruling: `内阁执行阶段异常 (${report.overall_status})，最高法院依法阻断进程。没有生成有效的执行产出。`,
        violation_type: ViolationType.DEVIATION_EXCEEDED,
        evidence: report.task_results.map(t => `[步骤 ${t.step_index}] ${t.status}: ${t.error ?? '无输出'}`),
        remediation: '建议立法分支细化执行步骤并重发法案',
        created_at: new Date()
      };
      await this.chiefJustice.issueJudgment(fakeVerdict);

      lifecycle.transition(BillState.UNCONSTITUTIONAL);
      await this._publishLifecycle(billId, "unconstitutional");
      lifecycle.transition(BillState.DRAFTING);
      return null;
    }

    lifecycle.transition(BillState.REVIEWING);
    await this._publishLifecycle(billId, "reviewing");

    const verdict = await this.chiefJustice.reviewResult(petition, report);
    await this.chiefJustice.issueJudgment(verdict);

    // Token 埋点④：司法审查阶段
    const judicialTokens = 2000;
    this._tokenCounters.judicial += judicialTokens;
    await this._publishTokenUsage(billId, 'judicial', judicialTokens, this._tokenCounters.judicial);

    if (!verdict.constitutional) {
      lifecycle.transition(BillState.UNCONSTITUTIONAL);
      await this._publishLifecycle(billId, "unconstitutional");
      lifecycle.transition(BillState.DRAFTING);
      return null;
    }

    lifecycle.transition(BillState.CONSTITUTIONAL);
    lifecycle.transition(BillState.DELIVERED);
    await this._publishLifecycle(billId, "delivered");

    return `法案 ${billId} 已交付。\n执行状态: ${report.overall_status}\n判决: ${verdict.ruling}\n总 Token 消耗: ${report.total_tokens_consumed}`;
  }

  private async _publishLifecycle(billId: string, state: string): Promise<void> {
    const event: BaseEvent = {
      timestamp: new Date(),
      source_agent: "government",
      action: EventAction.STATE_CHANGE,
      payload: { bill_id: billId, state },
      task_id: billId,
      emotion: EmotionType.NEUTRAL,
      intensity: 0,
    };
    await this.bus.publish("lifecycle", event);
  }

  private async _publishVotePassed(billId: string, ayes: number, nays: number, act?: any): Promise<void> {
    const event: VoteEvent = {
      timestamp: new Date(),
      source_agent: "speaker",
      ayes,
      nays,
      result: "passed",
      task_id: billId,
      action: EventAction.VOTE_PASSED,
      payload: act ? { act } : {},
      emotion: EmotionType.NEUTRAL,
      intensity: 0.8
    };
    await this.bus.publish("legislation", event);
  }

  private _forceVotePassed(voteResult: VoteResult): VoteResult & { _forced: boolean; _original_vote: { ayes: number; nays: number } } {
    // Bug 13 fix: 记录原始投票结果以供审计
    const originalAyes = voteResult.ayes;
    const originalNays = voteResult.nays;
    console.warn(
      `[CyberGovernment] ⚠️ 民主投票被否决 (ayes=${originalAyes}, nays=${originalNays})，` +
      `启动强制通过。原始投票结果已保留在 _original_vote 字段。`
    );

    return {
      proposal: voteResult.proposal,
      records: voteResult.records.map((r) => ({ voter_role: r.voter_role, vote: true })),
      ayes: voteResult.records.length,
      nays: 0,
      passed: true,
      _forced: true,
      _original_vote: { ayes: originalAyes, nays: originalNays },
    };
  }

  private async _publishTokenUsage(
    billId: string,
    branch: 'legislative' | 'executive' | 'judicial',
    tokensUsed: number,
    cumulative: number
  ): Promise<void> {
    const event: BaseEvent = {
      timestamp: new Date(),
      source_agent: 'government',
      action: EventAction.TOKEN_USAGE,
      payload: { branch, tokens_used: tokensUsed, cumulative },
      task_id: billId,
      emotion: EmotionType.NEUTRAL,
      intensity: 0,
    };
    await this.bus.publish('lifecycle', event);
  }
}
